src/components/RelicModal.jsx

```diff
@@ -331,174 +843,481 @@ export default function RelicModal(props) {
+                      onChange={(e) => {
+                        const stat = relicForm.getFieldValue('substatType0')
+                        if (stat) {
+                          if (stat == Constants.Stats.SPD) {
+                            setPrecision0(0)
+                            setHide0('none')
+                            setButton0low(2)
+                            setButton0mid(undefined)
+                            setButton0high(3)
+                            relicForm.setFieldValue('substatValue0', 0)
+                            setTrueStats({ substat0: 0 })
+                          } else {
+                            let low = SubStatValues[stat][5].low
+                            let mid = SubStatValues[stat][5].mid
+                            let high = SubStatValues[stat][5].high
+                            if ([Constants.Stats.HP, Constants.Stats.DEF, Constants.Stats.ATK].includes(stat)) {
+                              setPrecision0(0)
+                              low = Math.floor(low)
+                              mid = Math.floor(mid)
+                              high = Math.floor(high)
+                            } else {
+                              setPrecision0(1)
+                              low = Math.floor(10 * low) / 10
+                              mid = Math.floor(10 * mid) / 10
+                              high = Math.floor(10 * high) / 10
+                            }
+                            setButton0low(low)
+                            setButton0mid(mid)
+                            setButton0high(high)
+                            relicForm.setFieldValue('substatValue0', 0)
+                            setTrueStats({ substat0: 0 })
+                            setHide0('inline')
+                          }
+                        } else {
+                          relicForm.setFieldValue('substatValue0', undefined)
+                          setTrueStats({ substat0: 0 })
+                          setButton0low(undefined)
+                          setButton0mid(undefined)
+                          setButton0high(undefined)
+                          setHide0('inline')
+                        }
+                      }}
+                    />
+                  </Form.Item>
+
+                  <Form.Item size="default" name="substatValue0">
+                    <InputNumberStyled controls={false} precision={precision0} />
+                  </Form.Item>
+                </Flex>
+
+                <Flex gap={10}>
+                  <Form.Item size="default" name="substatType1">
+                    <Select
+                      showSearch
+                      allowClear
+                      style={{
+                        width: 200,
+                      }}
+                      placeholder="Substat"
+                      maxTagCount="responsive"
+                      options={substatOptions}
+                      onChange={(e) => {
```

Comment:
These 4 components also are duplicated and contain too much logic in the render code, these should probably be split out into a subcomponent with props, and reused